torch/testing/_internal/common_utils.py

```diff
@@ -803,204 +810,412 @@ def is_integral(self, dtype):
         dtypes = [x for x in get_all_dtypes() if x not in get_all_complex_dtypes()]
         return dtype in dtypes and not dtype.is_floating_point
 
-    # accepts tensors, dtypes, or np.ndarrays
-    def get_default_tolerance(self, a, b=None):
-        if b is None:
-            dtype = torch.float
-            if isinstance(a, torch.Tensor):
-                dtype = a.dtype
-            elif isinstance(a, torch.dtype):
-                dtype = a
-            elif TEST_NUMPY and isinstance(a, numpy.ndarray):
-                # Some tests call assertEqual with numpy Unicode.
-                if numpy.issubdtype(a.dtype, numpy.dtype('U')):
-                    dtype = torch.float
-                else:
-                    dtype = torch.from_numpy(a).dtype
-            if self.is_integral(dtype):
-                return (0, 0)
-            dtype = str(dtype).split('.')[-1]
-            return self.dtype_precisions.get(dtype, (self.precision, self.precision))
-
-        a_tol = self.get_default_tolerance(a)
-        b_tol = self.get_default_tolerance(b)
-        return (max(a_tol[0], b_tol[0]), max(a_tol[1], b_tol[1]))
-
-    def assertEqual(self, x, y, message='', *, atol=None, rtol=None, allow_inf=False, exact_dtype=None):
+    # Returns a dtype suitable for comparing tensors a and b.
+    # This closely follows NumPy's type promotion logic, except it includes
```

Comment:
why numpy / how does it differ from pytorch type promotion, and why not pytorch type promotion?

Reply:
PyTorch type promotion biases towards float types, so when comparing float16 x int64, for example, we would want to promote both to float16. NumPy, on the other hand, would want to promote both to double, which has a better chance of preserving the values in the int64 tensor. 

Preserving the values is a little academic today. It makes the debug messages nicer and avoids the unlikely scenario that an int64 tensor converted to float16 is incorrectly perceived as equal to the float16 tensor.

Reply:
For example:

```
torch.tensor((50000), dtype=torch.long).to(torch.float16)
: tensor(49984., dtype=torch.float16)
```

I wanted to lean towards a "future-proof" implementation that embodied the promised semantics as closely as possible. If we use PyTorch's type promotion (which is reasonable) than 50000 will compare equal to 49984 when comparing float16 x int64 tensors (in the future where we support the necessary functions on float16, anyway).

Reply:
I keep thinking we can do it in a followup PR, but at some point we should rename x or y, whichever is more appropriate/conventional or appears to be common usage in our tests, to `expected`.

I'm not sure how much that helps, but it seems like it would be atypical for a user to assert that their float16 is within a tight tolerance of a more precise int64 expected value. If they pass an int64 as `expected`, that may just have been the result of how we py-bind an integer literal, and I'd think we still want to favor the looser float16 tolerances. I guess this is just for the actual execution of the comparison op though and not the tolerance setting.

Either way, if all this code is really necessary, there might be some clearer way to explain it in the comments. I think if I came across this without context, I'd have a hard time understanding the intent well enough to correctly maintain it.

Reply:
I greatly expanded the comments to clarify. Renaming the args makes sense but it should probably be a follow-up so we don't continue to scope creep this pr. assert_allclose correctly has "actual" and "expected' but assertEqual does not.

Reply:
Oh, I forgot to clarify that in the case you described, where a float16 and int64 tensor are compared, then the float16 rtol and atol defaults are still used (if rtol and atol are not specified by the user). So the behavior you mention does occur.